ui/app/controllers/clients/index.js

```diff
@@ -25,12 +32,97 @@ export default Controller.extend(Sortable, Searchable, {
 
   searchProps: computed(() => ['id', 'name', 'datacenter']),
 
-  listToSort: alias('nodes'),
+  qpClass: '',
+  qpStatus: '',
+  qpDatacenter: '',
+  qpFlags: '',
+
+  selectionClass: selection('qpClass'),
+  selectionStatus: selection('qpStatus'),
+  selectionDatacenter: selection('qpDatacenter'),
+  selectionFlags: selection('qpFlags'),
+
+  optionsClass: computed('nodes.[]', function() {
+    const classes = Array.from(new Set(this.get('nodes').mapBy('nodeClass'))).compact();
+
+    // Remove any invalid node classes from the query param/selection
+    scheduleOnce('actions', () => {
+      this.set('qpClass', serialize(intersection(classes, this.get('selectionClass'))));
+    });
```

Comment:
This all seems correct, but I'm a little wary of side effects in computed properties. What do you think about moving these QP clean-ups into a method which is called from the route's `setupController` hook, after the model is set?

Reply:
I definitely like that better, but it won't capture the case where via blocking queries the underlying data changes and invalidates query params (e.g., clients change node class for some reason).

Although... maybe I could add some new hook to routes that have Watchers... 🤔

Reply:
Okay, I did some light exploration on this idea and I think it has legs. I'd prefer to do it in a follow up PR though.

Reply:
👍 I'd completely forgotten about blocking queries when I suggested it. Makes sense to explore post-merge.